### PR TITLE
Save state on the URL for filters and map zoom

### PR DESCRIPTION
**Pull request by melalj**
Demo: http://elalj.com/PCH/maptable-dev-38/#!mt-zoom=[3.7580909968560543,-1632.0471402331027,-257.11573666359783]

### CODE REVIEW

**Comment by melalj** [comment]
Related to #46

**Comment by Ths2-9Y-LqJt6** [comment]
The zoom seems to work - but not perfectly (different area is zoomed too).  I'll do some more testing.

The filter doesn't seem to work.  I added a filter on the example `index.html` for `country = 'it` (Italy) and my URL gets changed to `http://localhost:5000/#!mt-filters={%22country%22:[%22field%22,%22%22,%22ita%22]}!mt-zoom=[10,-2567.594276135132,-536.7695763949193]`.  However, loading that in a new browser (chromium or FF), the filter is not updated, only the zoom.

Finally - does this account for having multiple maptables on the DOM? I didn't see any DOM IDs in the URL to differentiate which maptable to update.

File: src/components/Filters.js [hunk 1]
```diff
@@ -109,6 +112,96 @@ export default class Filters {
...
+
+  restoreState() {
+    this.restoringState = true;
+    const params = document.location.href.split('!mt-filters=');
+    const defaultCriteria = (params[1]) ? params[1].split('!mt')[0] : null;
+    if (defaultCriteria) {
+      try {
+        this.setCriteria(JSON.parse(defaultCriteria));
+      } catch (e) {
+        console.log(`Maptable: Invalid URL State for mt-filters ${e.message}`);
```
Comment (melalj) [inline src/components/Filters.js]: `throw` is a blocking statement. And we don't want to block the JS execution if the URL state is wrong.

**Comment by Ths2-9Y-LqJt6** [comment]
We should also add a watcher for hashchange so that if you're on the page already and some one sends you a new view of it, the page updates.  something like [this](https://blog.plip.com/2015/04/25/named-anchor-navigation-using-javascript/).

**Comment by melalj** [comment]
Fixed deeplink when the window changes the size, escaping special characters on the URL: 42a6336

Added support of hashchange: c9c022e

File: src/components/Filters.js [hunk 2]
```diff
@@ -103,10 +110,103 @@ export default class Filters {
...
+      rowNodes[i].parentNode.removeChild(rowNodes[i]);
+    }
     this.criteria = [];
-    this.container.innerHTML = '';
-    this.refresh();
-    this.maptable.map.reset();
+    this.maptable.render();
+  }
+
+  exportCriteria() {
```
Comment (Ths2-9Y-LqJt6) [inline src/components/Filters.js]: for all the new methods like `exportCriteria()` here and others we've added, we should add JavaDoc blocks. I know a lot of the other functions/methods haven't been documented, but we should start now and slowly work toward 100% coverage!

**Comment by Ths2-9Y-LqJt6** [comment]
done with code review on this - on to QA!  normally should do it that in the other order though...sorry!

File: dist/maptable.js [hunk 3]
```diff
@@ -1021,7 +1172,7 @@ this.d3.maptable = (function () {
     }, {
       key: 'buildRow',
       value: function buildRow(filterName) {
-        var _this2 = this;
+        var _this3 = this;
```
Comment (melalj) [inline dist/maptable.js]: You shouldn't read this part. It's a auto-generated code by Babel ;)

File: dist/maptable.js [hunk 4]
```diff
@@ -2,6 +2,39 @@ this.d3 = this.d3 || {};
 this.d3.maptable = (function () {
   'use strict';
 
+  var babelHelpers = {};
```
Comment (melalj) [inline dist/maptable.js]: Sure, I generated the dist for test purposes.
This is a auto-generated code

File: dist/maptable.js (hunk 3 above)
Comment (Ths2-9Y-LqJt6) [inline dist/maptable.js]: ack!  so true.  i didn't see i'd wondered off into the `dist/` directory - sorry!

File: src/components/GeoMap.js [hunk 5]
```diff
@@ -104,6 +105,11 @@ export default class GeoMap {
     // Let's build things
     this.loadGeometries();
 
+    // Restore the state
+    this.restoreState();
+    window.addEventListener('hashchange', () => {
```
Comment (melalj) [inline src/components/GeoMap.js]: True.. Will refactor this

File: src/components/Filters.js (hunk 2 above)
Comment (melalj) [inline src/components/Filters.js]: Almost no method have the signature documentation. I prefer to have verbose naming of variables and functions instead to save time.
If you'd like to do it while discovering the code, you're welcome to open a PR for this ;)

**Comment by Ths2-9Y-LqJt6** [comment]
hrm - i still can't reliably do the following steps in FF:
1. start the dev environ with `gulp`
1. go to http://localhost:5000/
1. zoom in to italy
1. add a filter of "country = ita"
1. copy the URL
1. close the tab
1. open a new tab and paste the URL and hit return

expected: zoom and filters are reloaded 
actual: only zoom is restored

it still appears that when I copy the URL the `"`s get converted to `%22`s which makes the value for `mt-filters` get rejected.  so i paste this URL:
* http://localhost:5000/#!mt-filters={%22country%22:[%22field%22,%22%22,%22ita%22]}!mt-zoom=[10,-5322,-1457]
and I end up with this URL:
* http://localhost:5000/#!mt-zoom=[10,-5322,-1457]

File: src/components/GeoMap.js (hunk 5 above)
Comment (melalj) [inline src/components/GeoMap.js]: Fixed on 91e040f

File: src/components/GeoMap.js [hunk 6]
```diff
@@ -354,8 +362,71 @@ export default class GeoMap {
...
+    if (this.restoringState && this.options.map.saveState) return;
+    const encodedZoom = this.adaptToEncoded(this.scale, [this.transX, this.transY]);
+    const exportedCriteria = [encodedZoom.scale, encodedZoom.translate[0],
+      encodedZoom.translate[1]];
+    const params = document.location.href.split('!mt-zoom=');
+    const defaultZoom = (params[1]) ? params[1].split('!mt')[0] : null;
+    let newUrl = document.location.href.replace(`!mt-zoom=${defaultZoom}`, '');
+    if (this.scale !== 1) {
+      if (newUrl.indexOf('#') === -1) newUrl += '#';
+      newUrl += `!mt-zoom=${escape(JSON.stringify(exportedCriteria))}`;
```
Comment (melalj) [inline src/components/GeoMap.js]: Fixed on 5446b58

**Comment by melalj** [comment]
Sure thing. It's a trade off of time to invest to document the code vs adding features.
That should be maybe in a separate PR with consistent document in the code

**Comment by melalj** [comment]
It seems to be fixed with `encodeURIComponent` with 5446b58

**Pull request merged**